### PR TITLE
Improve UI Observables management

### PR DESCRIPTION
**Pull request by beorostica**
This is meant to solve UI issues #66 and #198 and organize the code related to the UI.

### CODE REVIEW

File: KomaMRICore/src/datatypes/Phantom.jl [hunk 1]
```diff
@@ -25,22 +27,22 @@ The Phantom struct.
...
 ```julia-repl
-julia> obj = Phantom()
+julia> obj = Phantom(x=[0.0])
 
 julia> obj.ρ
 ```
 """
  @with_kw mutable struct Phantom{T<:Real}
-    name::String = "spins"
+    name::String = "magnetizations"
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/Phantom.jl:36]: Edited on branch docs.

File: KomaMRICore/src/datatypes/Scanner.jl [hunk 2]
```diff
@@ -2,13 +2,14 @@
...
-The Scanner struct.
+The Scanner struct. It contains hardware limitations of the MRI resonator. It is an input
+for the simulation.
 
 # Arguments
 - `B0`: (`::Real`, `=1.5`, `[T]`) main magnetic field strength
 - `B1`: (`::Real`, `=10e-6`, `[T]`) maximum RF amplitude
 - `Gmax`: (`::Real`, `=60e-3`, `[T/m]`) maximum gradient amplitude
-- `Smax`: (`::Real`, `=500`, `[mT/m/ms]`) gradient maximum slew-rate
+- `Smax`: (`::Real`, `=500`, `[mT/m/ms]`) maximum gradient slew-rate
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/Scanner.jl:12]: Edited on branch docs.

File: KomaMRICore/src/datatypes/Sequence.jl [hunk 3]
```diff
@@ -32,60 +33,50 @@ mutable struct Sequence
...
-		GR.dur,
-		Dict()
-		)
+# Main Constructors
+function Sequence(GR)
+	M, N = size(GR)
+    gr = [i <= M ? GR[i,j] : Grad(0, 0) for i in 1:3, j in 1:N]
+    rf = reshape([RF(0, 0) for i in 1:N], 1, :)
+    adc = [ADC(0, 0) for _ = 1:N]
+	return Sequence(gr, rf, adc, GR.dur, Dict())
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/Sequence.jl:42]: Weird indeed. Edited on branch docs.

File: KomaMRICore/src/datatypes/sequence/ADC.jl [hunk 4]
```diff
@@ -3,26 +3,31 @@
...
-    T::Float64
-    delay::Float64
-    Δf::Float64
-    ϕ::Float64
+    N::Integer
+    T::Real
+    delay::Real
+    Δf::Real
+    ϕ::Real
     function ADC(N, T, delay, Δf, ϕ)
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/sequence/ADC.jl:31]: Maybe it is a good idea to do the same as in the phantom definition with a type T for the Float64 fieldnames (where T<:Number). For N define Integer supertype is more general. I reverted this in the docs branch.

File: KomaMRICore/src/datatypes/Sequence.jl [hunk 5]
```diff
@@ -4,20 +4,21 @@
     seq = Sequence(GR, RF)
     seq = Sequence(GR, RF, ADC)
     seq = Sequence(GR, RF, ADC, DUR)
-    seq = Sequence(GR::Array{Grad,1})
-    seq = Sequence(GR::Array{Grad,1}, RF::Array{RF,1})
-    seq = Sequence(GR::Array{Grad,1}, RF::Array{RF,1}, A::ADC, DUR, DEF)
+    seq = Sequence(GR::Vector{Grad})
+    seq = Sequence(GR::Vector{Grad}, RF::Vector{RF})
+    seq = Sequence(GR::Vector{Grad}, RF::Vector{RF}, A::ADC, DUR, DEF)
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/Sequence.jl:9]: In general, they are matrices. However there are created some constructors that accept vectors. I'm going to use the Array{<type>,1} notation for Vectors though.

File: KomaMRICore/src/datatypes/sequence/ADC.jl [hunk 6]
```diff
@@ -3,26 +3,31 @@
     adc = ADC(N, T, delay)
     adc = ADC(N, T, delay, Δf, ϕ)
 
-The ADC struct.
+The ADC struct. This is an event of an MRI sequence.
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/sequence/ADC.jl:6]: Edited.

File: KomaMRICore/src/datatypes/sequence/Grad.jl [hunk 7]
```diff
@@ -44,22 +44,29 @@ rotz(θ::Real) = [cos(θ) -sin(θ)	0;
...
-    grad = Grad(A, T, rise)
-    grad = Grad(A, T, rise, delay)
-    grad = Grad(A, T, rise, fall, delay)
+    gr = Grad(A, T)
+    gr = Grad(A, T, rise)
+    gr = Grad(A, T, rise, delay)
+    gr = Grad(A, T, rise, fall, delay)
 
-The Gradient struct.
+The Gradient struct. This is an event of an MRI sequence.
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/sequence/Grad.jl:52]: Edited.

File: KomaMRICore/src/datatypes/sequence/Grad.jl [hunk 8]
```diff
@@ -82,35 +89,33 @@ mutable struct Grad
...
 # Returns
-- `grad`: (`::Grad`) gradient struct
+- `gr`: (`::Grad`) gradient struct
 
 # Examples
 ```julia-repl
-julia> f1 = t -> sin(π*t / 0.8)
-
-julia> seq = Sequence([Grad(f1, 0.8)])
+julia> gr = Grad(t -> sin(π*t / 0.8), 0.8)
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/sequence/Grad.jl:110]: Edited.

File: KomaMRICore/src/datatypes/sequence/Grad.jl [hunk 9]
```diff
@@ -82,35 +89,33 @@ mutable struct Grad
 end
 
 """
-    grad = Grad(f::Function, T::Real, N::Int64; delay::Real)
+    gr = Grad(f::Function, T::Real, N::Integer; delay::Real)
 
-Generates an arbitrary gradient waveform defined by function `f` in the interval t ∈
-[0,`T`]. It uses `N` square gradients uniformly spaced in the interval.
+Generates an arbitrary gradient waveform defined by the function `f` in the interval t ∈
+[0,`T`]. The time separation between two consecutive samples is given by T/N.
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/sequence/Grad.jl:95]: My bad. Edited.

File: KomaMRICore/src/datatypes/sequence/RF.jl [hunk 10]
```diff
@@ -3,22 +3,29 @@
     rf = RF(A, T, Δf)
     rf = RF(A, T, Δf, delay)
 
-The RF struct.
+The RF struct. This is an event of an MRI sequence.
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/sequence/RF.jl:6]: Edited.

File: KomaMRICore/src/datatypes/sequence/RF.jl [hunk 11]
```diff
@@ -3,22 +3,29 @@
...
 
 # Arguments
-- `A`: (`::Complex{Int64}`, `[T]`) the amplitud-phase B1x + i B1y
-- `T`: (`::Int64`, [`s`]) the duration of the RF
-- `Δf`: (`::Float64`, [`Hz`]) the frequency offset of the RF
-- `delay`: (`::Float64`, [`s`]) the delay time of the RF
+- `A`: (`::Complex`, `[T]`) RF amplitud and phase. B1x + i B1y
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/sequence/RF.jl:9]: Edited.

File: KomaMRICore/src/datatypes/sequence/RF.jl [hunk 12]
```diff
@@ -3,22 +3,29 @@
...
+- `T`: (`::Real`, [`s`]) RF duration
+- `Δf`: (`::Real`, [`Hz`]) RF frequency dispracement from carrier
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/sequence/RF.jl:11]: Edited.

File: KomaMRICore/src/datatypes/simulation/DiscreteSequence.jl [hunk 13]
```diff
@@ -1,3 +1,22 @@
+"""
+    seqd = DiscreteSequence(Gx, Gy, Gz, B1, Δf, ADC, t, Δt)
+
+A discretized version of a Sequence struct, containing vectors for event amplitudes at
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/simulation/DiscreteSequence.jl:4]: Edited

File: KomaMRICore/src/datatypes/simulation/DiscreteSequence.jl [hunk 14]
```diff
@@ -50,12 +69,27 @@ is_GR_off(seq::DiscreteSequence) =  !is_GR_on(seq)
 is_RF_off(seq::DiscreteSequence) =  !is_RF_on(seq)
 is_ADC_off(seq::DiscreteSequence) = !is_ADC_on(seq)
 
-function discretize(seq::Sequence; simParams=default_sim_params())
-    t, Δt      = get_uniform_times(seq, simParams["Δt"]; Δt_rf=simParams["Δt_rf"])
+"""
+    seqd = discretize(seq::Sequence; sim_params=default_sim_params())
+
+This function returns a discretized Sequence struct with RF and gradient time refinements
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/simulation/DiscreteSequence.jl:75]: Done

File: KomaMRICore/src/io/Pulseq.jl [hunk 15]
```diff
@@ -278,10 +278,10 @@ end
 """
     seq = read_seq(filename)
 
-Returns the Sequence struct from a sequence file `.seq`.
+Returns the Sequence struct from a pulseq sequence file with `.seq` extension.
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/io/Pulseq.jl:281]: Edited

File: KomaMRICore/src/datatypes/sequence/Grad.jl [hunk 16]
```diff
@@ -44,22 +44,29 @@ rotz(θ::Real) = [cos(θ) -sin(θ)	0;
...
+    gr = Grad(A, T, rise, fall, delay)
 
-The Gradient struct.
+The Gradient struct. This is an event of an MRI sequence.
 
 # Arguments
-- `A`: (`::Float64`, `[T]`) amplitude of the gradient
-- `T`: (`::Float64`, `[s]`) duration of the flat-top
+- `A`: (`::Real`, `[T/m]`) amplitude of the gradient
+- `T`: (`::Real`, `[s]`) duration of the flat-top
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/sequence/Grad.jl:56]: It was an old mistake. Edited.

File: KomaMRICore/src/datatypes/Sequence.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (cncastillo) [reply KomaMRICore/src/datatypes/Sequence.jl:9]: Then put all the methods, the one that uses matrices and the one that uses Vectors.

File: KomaMRICore/src/io/ISMRMRD.jl [hunk 17]
```diff
@@ -1,36 +1,40 @@
...
-    raw_ismrmrd = signal_to_raw_data(signal, seq; phantom_name, sys, simParams)
+    raw = signal_to_raw_data(signal, seq; phantom_name, sys, sim_params)
 
-Transforms the raw signal into ISMRMRD format.
+Transforms the raw signal into a RawAcquisitionData struct (nearly equivalent to the ISMRMRD
+format) used for reconstruction with MRIReco.
 
 # Arguments
-- `signal`: (`::Vector{ComplexF64}`) raw signal
+- `signal`: (`::Vector{Complex}`) raw signal vector
```
[the remark this answers is not in the document]
Comment (cncastillo) [reply KomaMRICore/src/io/ISMRMRD.jl:8]: ??

File: KomaMRICore/src/datatypes/sequence/ADC.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (cncastillo) [reply KomaMRICore/src/datatypes/sequence/ADC.jl:31]: If it has the same performance it is fine.

File: KomaMRICore/src/datatypes/Sequence.jl [hunk 18]
```diff
@@ -32,60 +33,50 @@ mutable struct Sequence
...
+end
+function Sequence(GR::Array{Grad,1}, RF::Array{RF,1})
+    return Sequence(reshape(GR, :, 1), reshape(RF, 1, :), [ADC(0, 0) for i in 1:size(GR, 2)])
+end
+function Sequence(GR::Array{Grad,1}, RF::Array{RF,1}, A::ADC, DUR, DEF)
+    return Sequence(reshape(GR, :, 1), reshape(RF, 1, :), [A], [DUR], DEF)
+end
+function Sequence()
+    return Sequence([Grad(0, 0)])
+end
```
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/datatypes/Sequence.jl:79]: Done.

File: KomaMRICore/src/io/ISMRMRD.jl (hunk 17 above)
[the remark this answers is not in the document]
Comment (beorostica) [reply KomaMRICore/src/io/ISMRMRD.jl:8]: Edited.

**Pull request merged**